pkg/metrics/check_metrics.go

```diff
@@ -0,0 +1,158 @@
+	tlmCheckMetricsExpiredStateless = tlmCheckMetricsExpired.WithValues("false")
+	tlmCheckMetricsRemovedTotal     = tlmCheckMetricsRemoved.WithValues("sum")
+	tlmCheckMetricsRemovedStateful  = tlmCheckMetricsRemoved.WithValues("true")
+	tlmCheckMetricsRemovedStateless = tlmCheckMetricsRemoved.WithValues("false")
+)
+
+// CheckMetrics stores metrics for the check sampler.
+//
+// This is similar to ContextMetrics, but provides additional facility to remove metrics.
+type CheckMetrics struct {
+	expireMetrics bool
+	// additional time to keep stateful metrics in memory, after the context key has expired
+	statefulTimeout float64
+	metrics         ContextMetrics
+	deadlines       map[ckey.ContextKey]float64
+}
+
+// NewCheckMetrics returns new CheckMetrics instance.
+func NewCheckMetrics(expireMetrics bool, statefulTimeout time.Duration) CheckMetrics {
+	return CheckMetrics{
+		expireMetrics:   expireMetrics,
+		statefulTimeout: statefulTimeout.Seconds(),
+		metrics:         MakeContextMetrics(),
+		deadlines:       nil,
+	}
+}
+
+// AddSample adds a new sample to the metric with contextKey, initializing a new metric as necessary.
+//
+// If contextKey is scheduled for removal (see Remove), it will be unscheduled.
+//
+// See also ContextMetrics.AddSample().
+func (cm *CheckMetrics) AddSample(contextKey ckey.ContextKey, sample *MetricSample, timestamp float64, interval int64) error {
+	if cm.deadlines != nil {
+		delete(cm.deadlines, contextKey)
+	}
+	return cm.metrics.AddSample(contextKey, sample, timestamp, interval, checkMetricsAddSampleTelemetry)
+}
+
+// Remove deletes metrics associated with the given contextKeys.
+//
+// Some metrics can not be removed as soon as we stop receiving them (see
+// Metric.isStateful()). Stateful metrics will be kept around additional `cm.statefulTimeout`
+// time after this call, before ultimately removed. Call to AddSample will cancel delayed
+// removal.
+func (cm *CheckMetrics) Remove(contextKeys []ckey.ContextKey, timestamp float64) {
+	if !cm.expireMetrics {
+		return
+	}
+
+	expiredStateless := 0.0
+	expiredStateful := 0.0
+	for _, key := range contextKeys {
+		if m, ok := cm.metrics[key]; ok {
+			if m.isStateful() {
+				expiredStateful++
+				if cm.deadlines == nil {
+					cm.deadlines = make(map[ckey.ContextKey]float64)
+				}
+				cm.deadlines[key] = timestamp + cm.statefulTimeout
+			} else {
+				expiredStateless++
+				delete(cm.metrics, key)
+			}
+		}
+	}
+
+	tlmCheckMetricsExpiredTotal.Add(expiredStateless + expiredStateful)
+	tlmCheckMetricsExpiredStateful.Add(expiredStateful)
+	tlmCheckMetricsExpiredStateless.Add(expiredStateless)
+
+	tlmCheckMetricsRemovedTotal.Add(expiredStateless)
+	tlmCheckMetricsRemovedStateless.Add(expiredStateless)
+}
+
+// Flush flushes every metrics in the CheckMetrics (see ContextMetrics.Flush)
+func (cm *CheckMetrics) Flush(timestamp float64) ([]*Serie, map[ckey.ContextKey]error) {
+	return cm.metrics.Flush(timestamp)
+}
+
+// Cleanup removes expired stateful metrics.
```

Comment:
```suggestion
// Cleanup removes stateful metrics that have expired before the given timestamp.
```

This is a clever way to avoid timing in the tests :)